File: ios-native/ClubFuoco/Features/ClubDetail/DJPlayer.swift
import SwiftUI
import WebKit
import Observation

// App-level SoundCloud audio engine + our own gold UI.
//
// SoundCloud's player is a locked iframe, so we can't restyle it — but its JS
// Widget API lets us drive a HIDDEN widget (play/pause/seek + progress/title
// events) and build our own controls on top. The web view lives here in a
// singleton (attached off-screen to the key window), not inside any sheet, so
// audio KEEPS PLAYING when the DJ sheet closes and the user browses on. A
// persistent mini-bar (DJMiniBar) surfaces it app-wide.
//
// No background/lock-screen playback (no UIBackgroundModes: audio) — audio runs
// only while the app is foregrounded, by design.

@MainActor
@Observable
final class DJPlayer: NSObject {
    static let shared = DJPlayer()

    // Now-playing state (observed by the controls + mini-bar).
    private(set) var dj: FeaturedDJ?
    private(set) var isPlaying = false
    private(set) var isLoading = false
    private(set) var progress: Double = 0        // 0…1 of the current track
    private(set) var positionMs: Double = 0
    private(set) var durationMs: Double = 0
    private(set) var trackTitle = ""

    var isActive: Bool { dj != nil }

    private var webView: WKWebView?
    private var ready = false
    private var loadedProfile: String?           // avoid reloading the same DJ

    private nonisolated override init() { super.init() }

    // MARK: - Control

    /// Point the player at a DJ. Always loads that DJ (taking over any current
    /// playback); only starts playing when `autoplay` is true.
    func open(_ dj: FeaturedDJ, autoplay: Bool) {
        self.dj = dj
        guard let profile = Self.canonicalSoundCloud(dj.soundcloud)?.absoluteString else {
            reset(); return
        }
        if profile == loadedProfile, webView != nil {
            if autoplay { play() }
            return
        }
        trackTitle = ""; progress = 0; positionMs = 0; durationMs = 0
        isPlaying = false; isLoading = true
        loadedProfile = profile

        if webView == nil {
            buildWebView(initialProfile: profile, autoplay: autoplay)
        } else if ready {
            eval("scLoad('\(profile)', \(autoplay))")
        } else {
            // web view exists but not ready yet — swap the source and let READY fire.
            webView?.loadHTMLString(Self.html(profile: profile, autoplay: autoplay),
                                    baseURL: URL(string: "https://w.soundcloud.com"))
        }
    }

    func toggle() { isPlaying ? pause() : play() }
    func play()   { isLoading = true; eval("scPlay()") }
    func pause()  { eval("scPause()") }

    func seek(fraction: Double) {
        let f = min(1, max(0, fraction))
        progress = f
        positionMs = f * durationMs
        eval("scSeek(\(f))")
    }

    /// Stop and tear the player down (frees the web view).
    func close() {
        eval("scPause()")
        webView?.removeFromSuperview()
        webView?.configuration.userContentController.removeScriptMessageHandler(forName: "dj")
        webView = nil
        reset()
    }

    private func reset() {
        dj = nil; ready = false; loadedProfile = nil
        isPlaying = false; isLoading = false
        progress = 0; positionMs = 0; durationMs = 0; trackTitle = ""
    }

    // MARK: - Web view

    private func buildWebView(initialProfile: String, autoplay: Bool) {
        let cfg = WKWebViewConfiguration()
        cfg.allowsInlineMediaPlayback = true
        cfg.mediaTypesRequiringUserActionForPlayback = []   // the tap was the gesture
        cfg.userContentController.add(self, name: "dj")

        let web = WKWebView(frame: CGRect(x: 0, y: -2, width: 1, height: 1), configuration: cfg)
        web.isHidden = false          // .isHidden can suspend media — keep it "visible" but off-screen
        web.alpha = 0.01
        web.isUserInteractionEnabled = false
        // Attach off-screen to the key window so media is allowed to play.
        keyWindow()?.addSubview(web)
        webView = web
        web.loadHTMLString(Self.html(profile: initialProfile, autoplay: autoplay),
                           baseURL: URL(string: "https://w.soundcloud.com"))
    }

    private func eval(_ js: String) {
        webView?.evaluateJavaScript(js, completionHandler: nil)
    }

    private func keyWindow() -> UIWindow? {
        UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
            .first { $0.isKeyWindow }
    }

    // MARK: - Helpers

    /// RA stores "https://www.soundcloud.com/kink"; the widget resolver wants
    /// the canonical "https://soundcloud.com/kink".
    static func canonicalSoundCloud(_ raw: String?) -> URL? {
        guard var s = raw?.trimmingCharacters(in: .whitespaces), !s.isEmpty else { return nil }
        s = s.replacingOccurrences(of: "://www.soundcloud.com", with: "://soundcloud.com")
        return URL(string: s)
    }

    static func timeLabel(_ ms: Double) -> String {
        let total = Int((ms / 1000).rounded(.down))
        return String(format: "%d:%02d", total / 60, total % 60)
    }

    private static func html(profile: String, autoplay: Bool) -> String {
        let src = "https://w.soundcloud.com/player/?url=\(profile)"
            + "&auto_play=\(autoplay)&visual=false&hide_related=true&show_comments=false"
            + "&show_user=true&show_teaser=false&sharing=false&buying=false&download=false"
            + "&single_active=false&color=%23C09950"
        return """
        <!doctype html><html><head><meta name="viewport" content="width=device-width,initial-scale=1">
        <style>html,body{margin:0;height:100%;background:#fff;overflow:hidden}iframe{width:100%;height:100%;border:0}</style>
        </head><body>
        <iframe id="w" allow="autoplay" src="\(src)"></iframe>
        <script src="https://w.soundcloud.com/player/api.js"></script>
        <script>
        var w = SC.Widget(document.getElementById('w'));
        function post(m){ try{ window.webkit.messageHandlers.dj.postMessage(m); }catch(e){} }
        function meta(){ w.getCurrentSound(function(s){ if(s&&s.title) post({type:'title',title:s.title}); });
                         w.getDuration(function(d){ post({type:'duration',ms:d}); }); }
        w.bind(SC.Widget.Events.READY, function(){ post({type:'ready'}); meta(); });
        w.bind(SC.Widget.Events.PLAY, function(){ post({type:'play'}); meta(); });
        w.bind(SC.Widget.Events.PAUSE, function(){ post({type:'pause'}); });
        w.bind(SC.Widget.Events.FINISH, function(){ post({type:'finish'}); });
        w.bind(SC.Widget.Events.PLAY_PROGRESS, function(e){ post({type:'progress',ms:e.currentPosition,rel:e.relativePosition}); });
        function scPlay(){ w.play(); }
        function scPause(){ w.pause(); }
        function scSeek(f){ w.getDuration(function(d){ w.seekTo(Math.floor(d*f)); }); }
        function scLoad(url, ap){ w.load(url, { auto_play: ap, visual:false, hide_related:true,
            show_comments:false, show_user:true, sharing:false, buying:false, download:false,
            single_active:false, color:'%23C09950',
            callback: function(){ post({type:'loaded'}); meta(); if(ap){ w.play(); } } }); }
        </script></body></html>
        """
    }
}

extension DJPlayer: WKScriptMessageHandler {
    nonisolated func userContentController(_ controller: WKUserContentController,
                                           didReceive message: WKScriptMessage) {
        guard let body = message.body as? [String: Any],
              let type = body["type"] as? String else { return }
        Task { @MainActor in
            switch type {
            case "ready":
                self.ready = true
                if !self.isPlaying { self.isLoading = false }
            case "loaded":
                self.isLoading = false
            case "play":
                self.isPlaying = true; self.isLoading = false
            case "pause":
                self.isPlaying = false; self.isLoading = false
            case "finish":
                self.isPlaying = false; self.progress = 0; self.positionMs = 0
            case "title":
                if let t = body["title"] as? String { self.trackTitle = t }
            case "duration":
                if let d = body["ms"] as? Double { self.durationMs = d }
                else if let d = body["ms"] as? Int { self.durationMs = Double(d) }
            case "progress":
                if let rel = body["rel"] as? Double { self.progress = max(0, min(1, rel)) }
                if let pos = body["ms"] as? Double { self.positionMs = pos }
                else if let pos = body["ms"] as? Int { self.positionMs = Double(pos) }
            default: break
            }
        }
    }
}

// MARK: - In-sheet control (our own gold player)

/// Player card for the DJ sheet — the `.dj-player` box from the Club Fuoco DJ
/// Page design: accent play button, track title over a "SoundCloud preview"
/// overline, a waveform, and the running times, on a bordered surface card.
///
/// The waveform IS the scrubber. In the comp it is decoration; here the bars
/// carry the played fraction and take the drag, so the design's most
/// characteristic element does the work the old flat slider did rather than
/// sitting next to a duplicate control.
struct DJPlayerControl: View {
    @Environment(DJPlayer.self) private var player
    @Environment(LocaleStore.self) private var locale
    @State private var scrubbing: Double?          // fraction while dragging

    /// The comp's bar heights (32 bars, 28pt tall). A fixed figure, not random:
    /// it must not reshuffle on every redraw while audio is playing.
    private static let barHeights: [CGFloat] = [
        6, 11, 17, 9, 22, 14, 26, 10, 18, 24, 8, 15, 20, 12, 27, 9,
        16, 21, 11, 25, 13, 19, 7, 23, 17, 10, 14, 22, 9, 18, 12, 20,
    ]

    private var lineStrong: Color { Theme.fadedSand.opacity(0.34) }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 12) {
                Button { Haptics.tap(); player.toggle() } label: {
                    ZStack {
                        Circle().fill(Theme.ember).frame(width: 38, height: 38)
                        if player.isLoading {
                            ProgressView().tint(Theme.cream)
                        } else {
                            Image(systemName: player.isPlaying ? "pause.fill" : "play.fill")
                                .font(.system(size: 14))
                                .foregroundStyle(Theme.cream)
                                .offset(x: player.isPlaying ? 0 : 1.5)
                        }
                    }
                }
                .buttonStyle(.plain)

                VStack(alignment: .leading, spacing: 4) {
                    Text(player.trackTitle.isEmpty ? locale.t("dj.loadingTrack") : player.trackTitle)
                        .font(.cfSans(13.5, weight: .semibold))
                        .foregroundStyle(Theme.ink)
                        .lineLimit(1)
                    Text(locale.t("dj.soundcloudPreview").uppercased())
                        .font(.cfMono(9)).kerning(0.9)
                        .foregroundStyle(Theme.fadedSand)
                }
                Spacer(minLength: 0)
            }

            waveform.padding(.top, 14)

            HStack {
                Text(DJPlayer.timeLabel(player.positionMs))
                Spacer()
                Text(DJPlayer.timeLabel(player.durationMs))
            }
            .font(.cfMono(8.5)).kerning(0.5)
            .foregroundStyle(Theme.fadedSand)
            .padding(.top, 8)
        }
        .padding(.horizontal, 16).padding(.vertical, 14)
        .background(Theme.surface, in: .rect(cornerRadius: 16))
        .overlay(RoundedRectangle(cornerRadius: 16).stroke(lineStrong))
    }

    private var waveform: some View {
        GeometryReader { geo in
            let frac = scrubbing ?? player.progress
            let played = Int((Double(Self.barHeights.count) * frac).rounded())
            HStack(alignment: .bottom, spacing: 2) {
                ForEach(Array(Self.barHeights.enumerated()), id: \.offset) { i, h in
                    RoundedRectangle(cornerRadius: 1)
                        .fill(i < played ? Theme.ember : lineStrong)
                        .frame(height: h)
                }
            }
            .frame(width: geo.size.width, height: 28, alignment: .bottom)
            // The bars are 2pt apart, so hit-test the whole strip rather than
            // the glyphs — otherwise a scrub lands between bars and does nothing.
            .contentShape(Rectangle())
            .gesture(
                DragGesture(minimumDistance: 0)
                    .onChanged { v in
                        scrubbing = min(1, max(0, v.location.x / max(geo.size.width, 1)))
                    }
                    .onEnded { v in
                        let f = min(1, max(0, v.location.x / max(geo.size.width, 1)))
                        player.seek(fraction: f)
                        scrubbing = nil
                    }
            )
        }
        .frame(height: 28)
    }
}

File: android-native/app/src/main/java/com/clubfuoco/app/features/clubdetail/DjPlayer.kt
package com.clubfuoco.app.features.clubdetail

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import android.net.Uri
import android.view.ViewGroup
import android.webkit.JavascriptInterface
import android.webkit.WebSettings
import android.webkit.WebView
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableDoubleStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonPrimitive
import java.util.Locale

/**
 * The SoundCloud preview engine, and our own controls on top of it.
 *
 * SoundCloud's player is a locked iframe we cannot restyle, but its JS Widget
 * API lets us drive a HIDDEN widget — play, pause, seek, plus progress and title
 * events — and build the gold player around it.
 *
 * The web view lives in this singleton rather than in the DJ screen, so it
 * OUTLIVES that screen: the first open pays for the web view, the iframe and
 * SoundCloud's api.js, and every open after that is a profile swap on a warm
 * widget rather than a cold boot.
 *
 * Leaving the screen PAUSES rather than tears down. There is no app-wide mini
 * player, so audio must never outlive the screen the user can see — but the
 * machinery behind it should.
 *
 * Port of `DJPlayer.swift`.
 */
@SuppressLint("StaticFieldLeak")
object DjPlayer {

    // ── Now playing, observed by the controls ────────────────────────────────

    var raArtistId by mutableStateOf<String?>(null)
        private set
    var isPlaying by mutableStateOf(false)
        private set
    var isLoading by mutableStateOf(false)
        private set
    /** 0…1 of the current track. */
    var progress by mutableDoubleStateOf(0.0)
        private set
    var positionMs by mutableDoubleStateOf(0.0)
        private set
    var durationMs by mutableDoubleStateOf(0.0)
        private set
    var trackTitle by mutableStateOf("")
        private set

    /**
     * The widget answered with nothing playable — a dead profile, no public
     * tracks, or no network. Roughly one handle in ten is like this, so it is a
     * state the card states plainly rather than a transient it spins on.
     */
    var unavailable by mutableStateOf(false)
        private set

    // ── Machinery ────────────────────────────────────────────────────────────

    private var webView: WebView? = null
    private var ready = false
    private var loadedProfile: String? = null
    private var watchdog: Job? = null
    private val scope = CoroutineScope(Dispatchers.Main.immediate)
    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Longest a profile may sit loading before it is called unavailable. The
     * widget reports a bad profile itself within about half a second and gives
     * up on metadata by ~3.5s, so this only covers the page never coming up at
     * all — no network, or api.js blocked.
     */
    private const val LOAD_TIMEOUT_MS = 10_000L

    /**
     * Point the player at a DJ. Always loads that profile, taking over whatever
     * was playing; only starts playing when [autoplay].
     */
    fun open(context: Context, raArtistId: String, soundcloud: String?, autoplay: Boolean) {
        this.raArtistId = raArtistId

        val profile = canonicalSoundCloud(soundcloud)
        if (profile == null) {
            clear()
            unavailable = true
            return
        }

        if (profile == loadedProfile && webView != null) {
            // Same DJ, warm widget: nothing to reload.
            if (autoplay && !unavailable) play()
            return
        }

        trackTitle = ""
        progress = 0.0
        positionMs = 0.0
        durationMs = 0.0
        isPlaying = false
        isLoading = true
        unavailable = false
        loadedProfile = profile
        startWatchdog()

        val web = webView
        when {
            web == null -> build(context, profile, autoplay)
            ready -> eval("scLoad(${jsString(profile)}, $autoplay)")
            // Web view exists but has not reported READY yet — swap the source
            // and let READY fire against the new page.
            else -> web.loadDataWithBaseURL(
                BASE_URL, html(profile, autoplay), "text/html", "utf-8", null,
            )
        }
    }

    fun toggle() = if (isPlaying) pause() else play()

    fun play() {
        if (unavailable) return
        isLoading = true
        startWatchdog()
        eval("scPlay()")
    }

    fun pause() = eval("scPause()")

    fun seek(fraction: Double) {
        if (unavailable || durationMs <= 0) return
        val f = fraction.coerceIn(0.0, 1.0)
        progress = f
        positionMs = f * durationMs
        eval("scSeek($f)")
    }

    /**
     * Leaving the DJ screen: stop the audio, keep the widget warm.
     */
    fun suspend() {
        watchdog?.cancel()
        watchdog = null
        eval("scPause()")
        isPlaying = false
        isLoading = false
    }

    /** Stop and tear the player down, freeing the web view. */
    fun close() {
        eval("scPause()")
        webView?.let { web ->
            web.removeJavascriptInterface(BRIDGE)
            (web.parent as? ViewGroup)?.removeView(web)
            web.destroy()
        }
        webView = null
        clear()
    }

    private fun clear() {
        raArtistId = null
        ready = false
        loadedProfile = null
        isPlaying = false
        isLoading = false
        unavailable = false
        progress = 0.0
        positionMs = 0.0
        durationMs = 0.0
        trackTitle = ""
        watchdog?.cancel()
        watchdog = null
    }

    private fun startWatchdog() {
        watchdog?.cancel()
        watchdog = scope.launch {
            delay(LOAD_TIMEOUT_MS)
            if (isLoading) markUnavailable()
        }
    }

    private fun markUnavailable() {
        watchdog?.cancel()
        watchdog = null
        isLoading = false
        isPlaying = false
        // Only a profile that never produced a playable track is "unavailable";
        // one that already gave us a title is simply between events.
        if (trackTitle.isEmpty() || durationMs <= 0) unavailable = true
    }

    // ── Web view ─────────────────────────────────────────────────────────────

    @SuppressLint("SetJavaScriptEnabled")
    private fun build(context: Context, profile: String, autoplay: Boolean) {
        // Application context on the web view itself: this outlives any one
        // screen by design, and holding an Activity here would leak it.
        val web = WebView(context.applicationContext)
        web.settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            // The tap on our play button IS the user gesture — requiring a
            // second one inside the iframe would make the button do nothing.
            mediaPlaybackRequiresUserGesture = false
            mixedContentMode = WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE
        }
        web.addJavascriptInterface(Bridge, BRIDGE)
        attachOffScreen(context, web)
        webView = web
        web.loadDataWithBaseURL(BASE_URL, html(profile, autoplay), "text/html", "utf-8", null)
    }

    /**
     * Put the web view in the window, off screen.
     *
     * A WebView that is never attached can have its work throttled, which is
     * enough to stall the widget's bootstrap and lose its first metadata reply.
     * It is given a real player-sized box for the same reason — collapsing it to
     * a pixel throttles it just as hard. iOS does exactly this with the key
     * window; the note there was learned the hard way.
     *
     * Not fatal if there is no Activity to hand: the player simply runs
     * detached and may be slower to report in.
     */
    private fun attachOffScreen(context: Context, web: WebView) {
        val activity = generateSequence(context) { (it as? ContextWrapper)?.baseContext }
            .filterIsInstance<Activity>()
            .firstOrNull() ?: return
        val root = activity.window?.decorView as? ViewGroup ?: return

        val density = activity.resources.displayMetrics.density
        web.layoutParams = ViewGroup.LayoutParams(
            (320 * density).toInt(),
            (166 * density).toInt(),
        )
        // Off screen rather than GONE: a hidden view can have its media
        // suspended outright.
        web.translationY = -4000f
        web.alpha = 0.01f
        web.isEnabled = false
        runCatching { root.addView(web) }
    }

    private fun eval(js: String) {
        val web = webView ?: return
        // evaluateJavascript is main-thread only, and events can arrive on the
        // bridge's own thread.
        scope.launch { web.evaluateJavascript(js, null) }
    }

    private const val BRIDGE = "AndroidDj"
    private const val BASE_URL = "https://w.soundcloud.com"

    /** Receives the widget's events. Called on a WebView JS thread, not main. */
    private object Bridge {
        @JavascriptInterface
        fun onEvent(raw: String) {
            scope.launch { handle(raw) }
        }
    }

    private fun handle(raw: String) {
        val body = runCatching { json.parseToJsonElement(raw) as? JsonObject }.getOrNull() ?: return
        val type = body["type"]?.jsonPrimitive?.content ?: return
        val ms = body["ms"]?.jsonPrimitive?.doubleOrNull
        val rel = body["rel"]?.jsonPrimitive?.doubleOrNull

        when (type) {
            "ready" -> {
                ready = true
                if (!isPlaying) isLoading = false
            }
            "loaded" -> isLoading = false
            "play" -> {
                isPlaying = true
                isLoading = false
                unavailable = false
            }
            "pause" -> {
                isPlaying = false
                isLoading = false
            }
            "finish" -> {
                isPlaying = false
                progress = 0.0
                positionMs = 0.0
            }
            "title" -> body["title"]?.jsonPrimitive?.content?.takeIf { it.isNotEmpty() }?.let {
                trackTitle = it
                unavailable = false
            }
            "nometa", "error" -> markUnavailable()
            "duration" -> ms?.let { durationMs = it }
            "progress" -> {
                rel?.let { progress = it.coerceIn(0.0, 1.0) }
                ms?.let { positionMs = it }
            }
        }
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    /**
     * The catalogue stores "https://www.soundcloud.com/kink"; the widget
     * resolver wants the canonical "https://soundcloud.com/kink". A bare handle
     * is accepted too, since that is what a hand-entered value looks like.
     *
     * Anything that is not a soundcloud.com profile or track is REJECTED rather
     * than handed to the widget. A few rows hold a URL pasted into itself
     * ("soundcloud.com/https://raul-mezcolanza"), which parses happily and would
     * render a player that can never load.
     */
    fun canonicalSoundCloud(raw: String?): String? {
        var s = raw?.trim().orEmpty()
        if (s.isEmpty()) return null

        if (!s.lowercase().startsWith("http")) {
            if (s.contains("/") || s.contains(":")) return null
            s = "https://soundcloud.com/" + s.replace("@", "")
        }

        val uri = runCatching { Uri.parse(s) }.getOrNull() ?: return null
        val host = uri.host?.lowercase() ?: return null
        if (host != "soundcloud.com" && !host.endsWith(".soundcloud.com")) return null

        // A profile is one path segment, a track two, a set three
        // ("/kink/sets/live"). Deeper than that, or a segment carrying a scheme,
        // means the value is malformed rather than a page SoundCloud can serve.
        val segments = uri.path.orEmpty().split("/").filter { it.isNotEmpty() }
        if (segments.size !in 1..3) return null
        if (segments.any { it.contains(":") }) return null

        return "https://soundcloud.com/" + segments.joinToString("/")
    }

    fun timeLabel(ms: Double): String {
        val total = (ms / 1000).toInt().coerceAtLeast(0)
        return String.format(Locale.US, "%d:%02d", total / 60, total % 60)
    }

    private fun jsString(raw: String): String =
        "'" + raw.replace("\\", "\\\\").replace("'", "\\'") + "'"

    private fun html(profile: String, autoplay: Boolean): String {
        val encoded = Uri.encode(profile)
        val src = "https://w.soundcloud.com/player/?url=$encoded" +
            "&auto_play=$autoplay&visual=false&hide_related=true&show_comments=false" +
            "&show_user=true&show_teaser=false&sharing=false&buying=false&download=false" +
            "&single_active=false&color=%23C09950"

        // The metadata dance is the fiddly part and the comments are load-bearing
        // — see the iOS original, which learned it the hard way.
        return """
        <!doctype html><html><head><meta name="viewport" content="width=device-width,initial-scale=1">
        <style>html,body{margin:0;height:100%;background:#fff;overflow:hidden}iframe{width:100%;height:100%;border:0}</style>
        </head><body>
        <iframe id="w" allow="autoplay" src="$src"></iframe>
        <script src="https://w.soundcloud.com/player/api.js"></script>
        <script>
        var w = SC.Widget(document.getElementById('w'));
        function post(m){ try{ $BRIDGE.onEvent(JSON.stringify(m)); }catch(e){} }

        // A profile URL resolves to that artist's tracks, so the title and
        // duration have to be ASKED for. The widget only answers once its own
        // bootstrap has finished, which is often after READY fires — a single
        // ask on READY is why a player could sit on "Loading…" forever with a
        // live 0:00 / 0:00 under it.
        //
        // So ask repeatedly over a short window, then stop. A live profile
        // answers within ~100ms of READY; the later steps only cover a slow
        // bootstrap. Giving up at ~3.4s and saying so beats a longer wait,
        // because "no preview" is the honest answer by then.
        var haveTitle = false, haveDuration = false, gen = 0;
        var STEPS = [0, 300, 700, 1200, 2000, 3000];

        function meta(){
          w.getCurrentSound(function(s){
            if (s && s.title)    { haveTitle = true;    post({type:'title', title:s.title}); }
            if (s && s.duration) { haveDuration = true; post({type:'duration', ms:s.duration}); }
          });
          w.getDuration(function(d){
            if (d > 0) { haveDuration = true; post({type:'duration', ms:d}); }
          });
        }

        function done(){ return haveTitle && haveDuration; }

        // A new profile — or a track change inside one — invalidates what we
        // hold. `gen` retires the previous round's pending steps, so two loads
        // in quick succession cannot leave overlapping chains asking for the
        // metadata of a profile that is no longer on screen.
        function track(){
          haveTitle = false; haveDuration = false;
          var mine = ++gen;
          STEPS.forEach(function(delay, i){
            setTimeout(function(){
              if (mine !== gen || done()) return;
              try { meta(); } catch (e) {}
              if (i === STEPS.length - 1) {
                setTimeout(function(){
                  if (mine === gen && !done()) post({type:'nometa'});
                }, 400);
              }
            }, delay);
          });
        }

        w.bind(SC.Widget.Events.READY, function(){ post({type:'ready'}); track(); });
        w.bind(SC.Widget.Events.PLAY, function(){ post({type:'play'}); track(); });
        w.bind(SC.Widget.Events.PAUSE, function(){ post({type:'pause'}); });
        w.bind(SC.Widget.Events.FINISH, function(){ post({type:'finish'}); });
        w.bind(SC.Widget.Events.ERROR, function(){ post({type:'error'}); });
        w.bind(SC.Widget.Events.PLAY_PROGRESS, function(e){ post({type:'progress',ms:e.currentPosition,rel:e.relativePosition}); });

        function scPlay(){ w.play(); }
        function scPause(){ w.pause(); }
        function scSeek(f){ w.getDuration(function(d){ w.seekTo(Math.floor(d*f)); }); }
        function scLoad(url, ap){ w.load(url, { auto_play: ap, visual:false, hide_related:true,
            show_comments:false, show_user:true, sharing:false, buying:false, download:false,
            single_active:false, color:'#C09950',
            callback: function(){ post({type:'loaded'}); track(); if(ap){ w.play(); } } }); }
        </script></body></html>
        """.trimIndent()
    }
}
